### ISSUE TITLE
HTTP2: SocketsHttpHandler doesn't downgrade to HTTP/1.1 during Windows authentication

### ISSUE THREAD

**Issue by davidsh** [dotnet/runtime#1582]
When SocketsHttpHandler connects to a server using HTTP/2, it is unable to downgrade to HTTP/1.1 when the server requests Windows authentication (Negotiate or NTLM).  The net result is that 401 is returned to the client without even trying to reconnect (with HTTP/1.1) and send credentials.

```c#
// Repro will be posted later
```

**Comment by halter73** [comment]
This came up again. This time from a SignalR client user. See #45371. Prior to .NET 7, the SignalR client forced HTTP/1.1 but we started allowing HTTP/2 in .NET 7. We're considering backporting a fix for SignalR to force HTTP/1.1 if the customer sets `HttpConnectionOptions.UseDefaultCredentials` to true.

@karelz Even though we have a fix for SignalR, we can't help but think HttpClient should have avoided this for us. Do you think we could do something similar in .NET 8 at the SocketsHttpHandler level when UseDefaultCredentials is set? Of course, the handler could be smarter than SignalR and wait for a challenge before downgrading,

**Comment by karelz** [comment]
@wfurt @ManickaP @CarnaViire any thoughts on this?

**Comment by wfurt** [comment]
There are (at least) two distance parts IMHO. 
1) get parity with WinHttp - The existing retry logic will be problematic as this needs new connection (and possibly check for version policy). This can probably be done by refactoring the connection pool logic.
2) Browsers do have global cache and therefore they can remember state. `HttpClientHandler` is stateless and each request is processed independently. To fix that, we would need whole new concept in `SocketsHttpHandler`.  That may be useful for other reasons but it would be fundamental change IMHO.

While we _should_ do something I would keep it in `Future` for now given complexity and no up votes e.g. no real customer pain. The workaround is somewhat easy -> just set version to 1.1 for problematic sites.

**Comment by karelz** [comment]
Agreed with @wfurt. It seems it is still rather rare with a complex solution, so leaving it in Future sounds like the right option now.
@halter73 let us know if you disagree.

**Comment by MihaZupan** [comment]
The change I just merged #123827 works by remembering whether we see a session auth response on HTTP/2, and downgrading all future requests to that endpoint to use HTTP/1.1 (when allowed by version & policy). The initial H2 request that hits this will also be retried as long as it doesn't have a content.

@dotMorten Would you be able to test how this works in practice in your scenario once the preview with the change comes out (probably p4)?

**Comment by dotMorten** [comment]
@MihaZupan Yes I'll take a look then. Thanks!
